Use bookmarks created after the latest snapshot

0xFelix opened the pull request:
This commit changes syncoid's behavior so it is always looking for
a matching snapshot and a matching bookmark. If the bookmark was created
after the snapshot it is used instead. This allows replication when the
latest snapshot replicated was deleted on the source, a common snapshot
was found but rollback on the target is not allowed. The matching bookmark
is used instead for replication.

This fixes #602

0xFelix committed:
Use bookmarks created after the latest snapshot

This commit changes syncoid's behavior so it is always looking for
a matching snapshot and a matching bookmark. If the bookmark was created
after the snapshot it is used instead. This allows replication when the
latest snapshot replicated was deleted on the source, a common snapshot
was found but rollback on the target is not allowed. The matching bookmark
is used instead for replication.

This fixes https://github.com/jimsalterjrs/sanoid/issues/602

Signed-off-by: Felix Matouschek <felix@matouschek.org>

yhaenggi commented:
gentle bump

msladek commented:
would love to see this PR reviewed & merged @jimsalterjrs if time allows for it

0xFelix commented:
@jimsalterjrs @phreaker0 Ping

msladek commented:
@0xFelix sadly merge conflicts have creeped in again due to lots getting merged except this... I still rely on this PR in order to be able to use syncoid at all and would be willing to resolve the conflicts if I had write access to your repo. just let me know.

0xFelix commented:
@msladek Actually, I planned to pick up work on this and my other PR again soon, but help is always appreciated. Will contact you tomorrow.

0xFelix commented:
Rebased and working again.

Ping @phreaker0

0xFelix commented:
Also ping @jimsalterjrs

phreaker0 reviewed (changes requested):
this pr needs a rebase again, i just looked through and it looks good but I didn't test it.

I only have one concern, with this PR the bookmark list call is made for every dataset and may slow down the process for people which don't use the bookmark feature. Can you add a "no-bookmark" option to prevent the use of bookmarks?

0xFelix commented:
@bartmichu I'll see what I can do in the upcoming holidays.

0xFelix committed:
Add --no-bookmark flag

This prevents the use bookmarks when trying to find the latest common
snapshot. This forces a rollback when the latest snapshot on the source
was deleted but a common and older snapshot was found.

Signed-off-by: Felix Matouschek <felix@matouschek.org>

0xFelix commented:
@phreaker0 Rebased once more, please have a look

0xFelix commented:
Sorry for the noise, it is ready to review now

phreaker0 reviewed (approved):
Thank you very much and sorry again for the long delay reviewing this awesome feature!
